### PR TITLE
drivers: initial commit of generic AT parser module

### PR DESCRIPTION
**Pull request by kaspar030**
This PR adds a module for interacting with devices using AT commands.

### CODE REVIEW

**Comment by miri64** [comment]
@jia200x maybe useful for your https://github.com/RIOT-OS/RIOT/pull/5427

**Comment by vincent-d** [comment]
I started to work on prototyping with our modem some weeks ago, so I gave a try to your code.

It worked good once I understood I had to add a `\r` to each of my commands ;)

This is much more simpler that the system I used on a previous project which worked with a lot of callbacks. But here it's hard to deal with responses you do not expect.

I'll give some more try!

**Comment by jia200x** [comment]
@miri64 Indeed. Thanks for pointing me to this.

Cheers

**Comment by kaspar030** [comment]
> It worked good once I understood I had to add a \r to each of my commands ;)

oh, I hoped I had found a combination of line endings to send / line endings to expect that works. The two modems I tried accept a line with just ```\n```, but echo ```\r\n```. They don't like to receive crlf, though. Maybe we need to parameterize the newline behaviour somehow?

**Comment by kaspar030** [comment]
> This is much more simpler that the system I used on a previous project which worked with a lot of callbacks. But here it's hard to deal with responses you do not expect.

Very good point. I planned to add support for that later, once I come up with a process that works and doesn't suck. Suggestions welcome!

**Comment by vincent-d** [comment]
> Maybe we need to parameterize the newline behaviour somehow?

It would be handy, indeed.

I use a ublox modem, which needs a `\r` and echoes `\r\r\n`.

**Comment by aabadie** [comment]
Any update here ? @kaspar030

**Comment by vincent-d** [comment]
I gave another try and proposed some improvements in https://github.com/kaspar030/RIOT/pull/27

File: drivers/at/at.c [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+    unsigned cmdlen = strlen(command);
+
+    uart_write(dev->uart, (const uint8_t *)command, cmdlen);
+    uart_write(dev->uart, (const uint8_t *)AT_END_OF_LINE, sizeof(AT_END_OF_LINE) - 1);
+
+    if (at_expect_bytes(dev, command, cmdlen, timeout)) {
+        return -1;
+    }
+
+    if (at_expect_bytes(dev, AT_END_OF_LINE "\r\n", sizeof(AT_END_OF_LINE) + 1, timeout)) {
```
Comment (haukepetersen) [inline drivers/at/at.c]: shouldn't this be `AT_END_OF_LINE "\n"` as second parameter?

Comment (vincent-d) [inline drivers/at/at.c]: On my modem (and I think it works the same for others as well), if you send `AT\r` you get as a response:
`AT\r` <- echo
`\r\nOK\r\n`

Comment (haukepetersen) [inline drivers/at/at.c]: maybe I missunderstand the line above, but `AT_END_OF_LINE` is
```c
#define AT_END_OF_LINE "\r"
```
So the line above does make a string concatenation, leading to `\r\r\n`, right?!

Comment (vincent-d) [inline drivers/at/at.c]: Yes

Comment (haukepetersen) [inline drivers/at/at.c]: And this is wanted? I don't quite see it, but if it is intended then never mind...

Comment (vincent-d) [inline drivers/at/at.c]: Yes, the `AT_END_OF_LINE` is expected after the command, the `\r\n` before the first response line. 

It seems that some modems can use different end of line character than `\r`, that's why it's configurable.

Comment (haukepetersen) [inline drivers/at/at.c]: alright, I see.

Comment (haukepetersen) [inline drivers/at/at.c]: but just to make sure: wouldn't it make sense then, to make the characters that are expected before the first response line also configurable?

Comment (vincent-d) [inline drivers/at/at.c]: That would not hurt, indeed

File: drivers/include/at.h [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+ *
+ * A possible empty line will be skipped.
+ *
+ * @param[in]   dev         device to operate on
+ * @param[in]   command     command to send
+ * @param[out]  resp_buf    buffer for storing response
+ * @param[in]   len         len of @p buffer
+ * @param[in]   timeout     timeout (in usec)
+ *
+ * @returns     lenght of response on success
```
Comment (aabadie) [inline drivers/include/at.h]: before I forget, there's a typo: s/lenght/length/. I saw it in other places.

**Comment by keestux** [comment]
@aabadie Sorry, I don't have a LoRaBee at hand. I _do_ have sodaq ONE and ExpLoRer board with the same RN2483 device but unfortunately I can't connect my Atmel ICE to these boards (so I can't load RIOT programs). :-(

File: drivers/include/at.h [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+ * @returns     <0 on error
+ */
+ssize_t at_send_cmd_get_lines(at_dev_t *dev, const char *command, char *resp_buf, size_t len, uint32_t timeout);
+
+/**
+ * @brief   Expect bytes from device
+ *
+ * @param[in]   dev     device to operate on
+ * @param[in]   bytes   buffer containing bytes to expect
+ * @param[in]   len     number of bytes to expect
```
Comment (keestux) [inline drivers/include/at.h]: Why do we need 'len'? The length of 'bytes' should be leading, otherwise you need a sanity check in the function.

File: drivers/at/at.c [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+}
+
+ssize_t at_readline(at_dev_t *dev, char *resp_buf, size_t len, uint32_t timeout)
+{
+    ssize_t res = -1;
+    char *resp_pos = resp_buf;
+
+    memset(resp_buf, 0, len);
+
+    while (len) {
```
Comment (keestux) [inline drivers/at/at.c]: What happens if the response string is longer than 'len' and the while loop is ended. If I understand this code well enough then I'm afraid there is no NUL byte terminating the string.

File: drivers/include/at.h [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+ * @defgroup    drivers_at AT (Hayes) command set library
+ * @ingroup     drivers
+ * @brief       AT (Hayes) command set library
+ *
+ * This module provides functions to interact with devices using AT commands.
+ *
+ * Most functions compare the bytes echoed by the device with what they
+ * intended to send, and bail out if there's no match.
+ *
+ * Furthermore, the library tries to copy with difficulties regarding different
```
Comment (smlng) [inline drivers/include/at.h]: /copy/cope/ ?

File: drivers/include/at.h (hunk 3 above)
Comment (kaspar030) [reply drivers/include/at.h]: I don't quite get the question. ```len``` specifies the nr of bytes in ```bytes```. Maybe the docs are misleading?

File: drivers/at/at.c (hunk 4 above)
Comment (kaspar030) [reply drivers/at/at.c]: That's what the memset in line 210 should take care of?

File: drivers/include/at.h (hunk 5 above)
Comment (kaspar030) [reply drivers/include/at.h]: fixed

File: drivers/include/at.h (hunk 2 above)
Comment (kaspar030) [reply drivers/include/at.h]: fixed two times int at.h.

**Comment by kaspar030** [comment]
> Do you think it could be even more generalized to any string command controlled devices ? (not only AT).

It probably can even be used as is for devices not speaking Hayes. The actual AT commands are all built on top of ```at_expect_bytes()```, ```at_readline()``` and ```at_send_bytes()```.

File: drivers/at/at.c (hunk 4 above)
Comment (keestux) [reply drivers/at/at.c]: No, not quite. Say there is a 'len' of 5 and on the input comes "abcdefg\r\n". The while loop ends as soon as "abcde" was read and resp_buf is completely filled up (no NUL byte anymore). No?

File: drivers/include/at.h (hunk 3 above)
Comment (keestux) [reply drivers/include/at.h]: Why not rely on strlen(bytes) instead of having to pass a string and a length? In the implementation there is a while loop for 'len', but there is no sanity check whether there are that many chars in 'bytes'.

Without 'len' the usage could become as simple as:

    if (at_expect_bytes(dev, AT_END_OF_LINE "\r\n", timeout)) {

Bonus: in this case getting rid of the confusing sizeof()+1

Comment (kaspar030) [reply drivers/include/at.h]: > Why not rely on strlen(bytes) instead of having to pass a string and a length?

the function is mostly used by other functions (quick grep counts five times). relying on strlen would lead to counting strings multiple times, for strings that are often static. having ```len``` as parameter allows the API user to decide whether to use strlen or not...



> there is no sanity check whether there are that many chars in 'bytes'

File: drivers/at/at.c (hunk 4 above)
Comment (kaspar030) [reply drivers/at/at.c]: Ah, right. I guess ```-ENOSPC``` would be the desired result in that case?

File: drivers/include/at.h (hunk 3 above)
Comment (keestux) [reply drivers/include/at.h]: I'm not impressed by that argument.
You can change the while condition in at_expect_bytes() to

    while (*bytes != 0) {

and then you won't need to do a strlen()

File: drivers/at/at.c (hunk 4 above)
Comment (kaspar030) [reply drivers/at/at.c]: Looking at the code again, if no full line could be read, the result of the function is -1. Shouldn't that be enough? Anyhow, I've added a condition setting the first byte of resp_buf to the null byte in that case.

File: drivers/include/at.h (hunk 3 above)
Comment (kaspar030) [reply drivers/include/at.h]: That actually seems like a good idea. done.

**Comment by kaspar030** [comment]
Can we get forward with this? *ping* @haukepetersen @aabadie @keestux

File: drivers/at/at.c (hunk 4 above)
Comment (keestux) [reply drivers/at/at.c]: I still don't see how the string is terminated with a nul-byte. In my example above resp_buf contains the 5 characters a, b, c, d and e. The return value is 5. Maybe you say that the caller is responsible to not treat the result as a nul-terminated string?

File: drivers/Makefile.dep [hunk 6]
```diff
@@ -4,6 +4,11 @@ ifneq (,$(filter adxl345,$(USEMODULE)))
     FEATURES_REQUIRED += periph_i2c
 endif
 
+ifneq (,$(filter at,$(USEMODULE)))
+    USEMODULE += fmt
```
Comment (aabadie) [inline drivers/Makefile.dep]: indent should be 2 spaces

**Comment by aabadie** [comment]
I'll need this PR soon for a new radio driver

File: drivers/include/at.h [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+ * @brief   Initialize AT device struct
+ *
+ * @param[in]   dev         struct to initialize
+ * @param[in]   uart        UART the device is connected to
+ * @param[in]   baudrate    baudrate of the device
+ * @param[in]   buf         input buffer
+ * @param[in]   bufsize     size of @p buf
+ *
+ * @returns     0 on success
+ * @returns     <0 otherwise
```
Comment (aabadie) [inline drivers/include/at.h]: Using an anonymous enum for return codes would improve the API of this library.

Comment (aabadie) [inline drivers/include/at.h]: Using an anonymous enum for return codes would improve the API of this library.

File: drivers/include/at.h [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+ *
+ * @returns     0 when device answers "OK"
+ * @returns     <0 otherwise
+ */
+int at_send_cmd_wait_ok(at_dev_t *dev, const char *command, uint32_t timeout);
+
+/**
+ * @brief   send AT command, wait for a prompt
+ *
+ * This function will send the supplied @p command, then wait for the prompt (>)
```
Comment (aabadie) [inline drivers/include/at.h]: "This functions sends ..., then waits for ...". Prefer present instead of future (you did it like this above). Same in the documentations below.

File: drivers/at/at.c (hunk 4 above)
Comment (kaspar030) [reply drivers/at/at.c]: The return variable ```res```, which defaults to ```-1```, is only set to the number of bytes read if a full line could be read. In your example (```[ a, b, c, d, e ]```), the while loop would end with ```res == -1``` and thus set ```*resp_buf = '\0'```.

File: drivers/Makefile.dep (hunk 6 above)
Comment (kaspar030) [reply drivers/Makefile.dep]: fixed

File: drivers/at/at.c [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+int at_expect_bytes(at_dev_t *dev, const char *bytes, uint32_t timeout)
+{
+    while (*bytes) {
+        char c;
+        int res;
+        if ((res = isrpipe_read_timeout(&dev->isrpipe, &c, 1, timeout)) == 1) {
+            if (AT_PRINT_INCOMING) {
+                print(&c, 1);
+            }
+            if (!(c == *bytes++)) {
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply drivers/at/at.c]: yup

File: drivers/at/at.c [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+    if (res == 0) {
+        /* skip possible empty line */
+        res = at_readline(dev, resp_buf, len, timeout);
+    }
+
+out:
+    return res;
+}
+
+ssize_t at_send_cmd_get_lines(at_dev_t *dev, const char *command, char *resp_buf, size_t len, uint32_t timeout)
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply drivers/at/at.c]: done

File: drivers/at/at.c [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+    memset(resp_buf, '\0', len);
+
+    at_drain(dev);
+
+    res = at_send_cmd(dev, command, timeout);
+    if (res) {
+        goto out;
+    }
+
+    while(1) {
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply drivers/at/at.c]: yup

File: drivers/include/at.h (hunk 7 above)
Comment (kaspar030) [reply drivers/include/at.h]: not in this PR

File: drivers/include/at.h (hunk 8 above)
Comment (kaspar030) [reply drivers/include/at.h]: done and fixed some more style issues in the doxygen

**Comment by kaspar030** [comment]
- rebased
- addressed @aabadie's comments
- renamed ```AT_END_OF_LINE``` -> ```AT_EOL```, added ```AT_EOL_LEN``` define

File: drivers/at/at.c (hunk 4 above)
Comment (kaspar030) [reply drivers/at/at.c]: @keestux Does this address your concerns?

Comment (keestux) [reply drivers/at/at.c]: I can see that there can be no unterminated string, even when a longer line comes in. Good.

On the other hand, you take drastic measures to throw away the whole input (instead of truncating).

If this is what we want then OK.
BTW. The doc says "<0 on error", which is either a timeout (-ETIMEDOUT) or a too long input line (-1). Perhaps you want to put this in the function description (at.h).

**Comment by vincent-d** [comment]
Could we move forward on this one? I have some work pending waiting for this. @aabadie @keestux @haukepetersen ?

**Comment by keestux** [comment]
Since you asked ... I went over `at.c/at.h` once more and have the following findings. Sorry for the long message.

- Some devices (for example UBlox Cellular Modules, and SIM800/SIM900) have "Unsolicited Response Codes". These URC's can come in at any moment. That would cause problems if that happens between `uart_write` and `at_expect_bytes.` So, it is debatable whether the `at_expect_bytes` calls in `at_send_cmd` are a good idea. I understand that we want the at_expect_bytes to tightly synchronize the character we send with the reply from the modem. But maybe we need to do that character by character.

- The timeout in `at_drain` a hard coded number. It deserves a comment explaining why it is 10ms. No?

- Why use `unsigned` instead of `size_t?` (2x)

- Besides `+CME ERROR`, there is also `+CMS ERROR`. (In UBlox and SIM, for SMS errors)

- In `at_send_cmd_get_lines` the `out` label is used only once. Either do a simple return here, or do `goto out` in the remainder of the function.

- In `at_readline` there is a `goto out`, which could easily be replaced by a `break`, and then drop the label `out`.

- In `at_send_cmd_wait_prompt` there almost a complete repetition of `at_send_cmd`. Why is that? The only difference is the following. (If that's intended, please add a comment. If not, please replace by a call to `at_send_cmd`)

This

    if (at_expect_bytes(dev, AT_EOL "\r\n", timeout)) {

versus this

    if (at_expect_bytes(dev, AT_EOL "\n", timeout)) {

- A remark about the comment in at.h

    It usually sends "< command>< CR>", but expects "< command>\LF\CR" as echo.

- Notice that in `at_send_cmd` the code is expecting "< command>< CR>< CR>< LF>" (see also my remark about code duplication). Perhaps we need two macros, one for the < CR> to send a command and one for < CR>< LF> as the end-of-line.

BTW. I did all this just by code inspection, using my knowledge of writing the GPRSbee and Sodaq_3Gbee Arduino libraries. I did not actually test anything with RIOT, sorry.

**Comment by kaspar030** [comment]
> Some devices (for example UBlox Cellular Modules, and SIM800/SIM900) have "Unsolicited Response Codes". These URC's can come in at any moment. That would cause problems if that happens between uart_write and at_expect_bytes. So, it is debatable whether the at_expect_bytes calls in at_send_cmd are a good idea. I understand that we want the at_expect_bytes to tightly synchronize the character we send with the reply from the modem. But maybe we need to do that character by character.

URCs cannot come at any moment, AFAIK, they cannot arrive after AT has been sent and the corresponding command has been finished?

There's #8542 which might be a direction to go.
My hunch is that we need to have a different approach to URCs. Anyhow, I suggest getting this PR merged and improving later.

**Comment by kaspar030** [comment]
> The timeout in at_drain a hard coded number. It deserves a comment explaining why it is 10ms. No?

Fixed.

> Why use unsigned instead of size_t? (2x)

In at_send_cmd() and it's prompting brother? fixed

> Besides +CME ERROR, there is also +CMS ERROR. (In UBlox and SIM, for SMS errors)

Added.

> In at_send_cmd_get_lines the out label is used only once. Either do a simple return here, or do goto out in the remainder of the function.
> 
> In at_readline there is a goto out, which could easily be replaced by a break, and then drop the label out.

Do those once-used labels hurt a lot?

> In at_send_cmd_wait_prompt there almost a complete repetition of at_send_cmd. Why is that? The only difference is the following. (If that's intended, please add a comment. If not, please replace by a call to at_send_cmd)

I merged a PR to this PR, which introduced the extra function. Should've seen.
Anyhow, I rewrote the function so it now calls at_send_cmd(). @vincent-d does it still work?

**Comment by kaspar030** [comment]
Honestly I don't want to spend too much time on this. I suggest we either merge it (possibly marking it as "experimental"), or I keep it in a seperate repo.

File: drivers/at/at.c [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+
+out:
+    return res;
+}
+
+int at_send_cmd_wait_prompt(at_dev_t *dev, const char *command, uint32_t timeout)
+{
+    at_drain(dev);
+
+    int res = at_send_cmd(dev, command, timeout);
```
Comment (vincent-d) [inline drivers/at/at.c]: I didn't re-test yet, but IIRC, this won't work anymore.
Because in case of the prompt, the "\r\n" is not sent back.

In 'normal' case: we send `command\r\n` and we receive `command\r\n\r\nOK\r\n`. In case of a prompt, if we send `command\r\n` we receive `command\r\n>`. That's why I added a new function instead of reusing `at_send_cmd`.

**Comment by vincent-d** [comment]
> does it still work?

See my inline response

>  suggest we either merge it (possibly marking it as "experimental")

+1, there is for sure room for improvements, but I'd like this to be merge to be able to move forward on related topics.

File: drivers/at/at.c (hunk 12 above)
Comment (keestux) [reply drivers/at/at.c]: I thought we just send `command\r` (without the `\n`)

**Comment by keestux** [comment]
I agree with @kaspar030 , let's merge it.

**Comment by aabadie** [comment]
> I suggest we either merge it (possibly marking it as "experimental")

Like others, I also agree. From a reviewer point of view, e.g. skipping AT parser implementation details discussed above, the main problem with this PR is the missing test application. It would be great to provide a simple with a shell to send commands.
Otherwise, this module is not built by the CI.

File: drivers/at/at.c (hunk 12 above)
Comment (kaspar030) [reply drivers/at/at.c]: confusing at least ;)
Anyhow, I reverted to the working version. Let's optimize later.

**Comment by vincent-d** [comment]
monthly wake-up ;)

@kaspar030 do you think you can make it for the release?

File: tests/at/main.c [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+        puts("Please enter a command");
+        return 1;
+    }
+
+    if (at_send_cmd_get_resp(&at_dev, argv[1], resp, sizeof(resp), 10 * US_PER_SEC) > 0) {
+        puts("Response:");
+        puts(resp);
+    }
+    else {
+        puts("Error");
```
Comment (aabadie) [inline tests/at/main.c]: return 1 after ?

File: tests/at/main.c [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+
+static int send(int argc, char **argv)
+{
+    if (argc < 2) {
+        puts("Please enter a command");
+        return 1;
+    }
+
+    if (at_send_cmd_get_resp(&at_dev, argv[1], resp, sizeof(resp), 10 * US_PER_SEC) > 0) {
+        puts("Response:");
```
Comment (aabadie) [inline tests/at/main.c]: Why not use `printf("Response: %s\n", resp);` instead ? (a `\n` could be added after `:` to be strictly identical)

File: tests/at/main.c [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+#ifndef BAUDRATE_AT
+#define BAUDRATE_AT 115200
+#endif
+
+static int init(int argc, char **argv)
+{
+    (void)argc;
+    (void)argv;
+
+    at_dev_init(&at_dev, UART_AT, BAUDRATE_AT, buf, sizeof(buf));
```
Comment (aabadie) [inline tests/at/main.c]: I think, it would be more flexible to pass the UART and baudrate to the init function.

**Comment by aabadie** [comment]
@kaspar030, if you don't mind I'll address the comments I raised.

File: drivers/at/at.c [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+}
+
+int at_send_cmd(at_dev_t *dev, const char *command, uint32_t timeout)
+{
+    size_t cmdlen = strlen(command);
+
+    uart_write(dev->uart, (const uint8_t *)command, cmdlen);
+    uart_write(dev->uart, (const uint8_t *)AT_SEND_EOL, AT_SEND_EOL_LEN);
+
+    if (AT_SEND_ECHO) {
```
Comment (aabadie) [inline drivers/at/at.c]: @vincent-d @kaspar030, here is the change that allows to skip the echo check at build time.

**Comment by vincent-d** [comment]
> What about adding a macro like AT_END_OF_RESP

I would suggest to squash and merge as-is and improve in a follow-up PR.
I also have a couple of improvements I'd like to merge later on.

@kaspar030 could you squash?

**Comment by aabadie** [comment]
> we should fix Murdock

done! (sorry)

File: cpu/atmega_common/periph/i2c.c [hunk 17]
```diff
@@ -32,9 +32,6 @@
 #include "debug.h"
 #define ENABLE_DEBUG      (0)
 
-/* guard file in case no I2C device is defined */
```
Comment (aabadie) [inline cpu/atmega_common/periph/i2c.c]: This looks unrelated, weird.

Comment (kaspar030) [reply cpu/atmega_common/periph/i2c.c]: yes, no idea how that commit sneaked in. I squashed it out.

**Comment by aabadie** [comment]
@vincent-d, ok to merge ? I think we can safely dismiss @haukepetersen's review

**Comment by aabadie** [comment]
Cool !

**Comment by vincent-d** [comment]
Great team work! Thanks all!
